src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.ReadWrite.cs

```diff
@@ -0,0 +1,431 @@
+                        Input.Writer.Commit();
+                        _inputHandle.Dispose();
+                    }
+
+                    // Flush the read data for the Input Pipe writer
+                    var flushResult = await Input.Writer.FlushAsync();
+
+                    // If the pipe was closed, we are done reading, 
+                    if (flushResult.IsCompleted || flushResult.IsCancelled)
+                    {
+                        _doneReading = true;
+                    }
+                }
+
+                // Complete the input writer as we are done reading the request body.
+                Input.Writer.Complete();
+            }
+            catch (Exception ex)
+            {
+                Output.Reader.Complete(ex);
+                Input.Writer.Complete(ex);
+            }
+        }
+
+        private async Task WriteLoopAsync()
+        {
+            try
+            {
+                while (true)
+                {
+                    // Reading is done, so we will await all reads from the output pipe
+                    var readResult = await Output.Reader.ReadAsync();
+
+                    // Get data from pipe
+                    var buffer = readResult.Buffer;
+                    var consumed = buffer.End;
+
+                    try
+                    {
+                        // If the output pipe is termniated, cancel the request.
+                        if (readResult.IsCancelled)
+                        {
+                            break;
+                        }
+
+                        if (!buffer.IsEmpty)
+                        {
+                            // Write to IIS buffers
+                            // Guaranteed to write the entire buffer to IIS
+                            await WriteToIISAsync(buffer);
+                        }
+                        else if (readResult.IsCompleted)
+                        {
+                            break;
+                        }
+                        else
+                        {
+                            // Flush of zero bytes will 
+                            await FlushToIISAsync();
+                        }
+                    }
+                    finally
+                    {
+                        // Always Advance the data pointer to the end of the buffer.
+                        Output.Reader.AdvanceTo(buffer.End);
+                    }
+
+                    // Check if there was an upgrade. If there is, we will replace the request and response bodies with
+                    // two seperate loops
+                    if (_upgradeTcs?.TrySetResult(null) == true)
+                    {
+                        // The Input pipe has already been completed at this point
+                        // Create a new Input pipe for websockets
+                        Input = new Pipe(new PipeOptions(_memoryPool, readerScheduler: PipeScheduler.ThreadPool, minimumSegmentSize: MinAllocBufferSize));
```

Comment:
Ugh, this is pretty gross. When would this ever happen? How does the input pipe close for an upgraded request? Is this an artifact of how the code is written or can this actually happen?